### DISCUSSION TITLE
Gesture Typing

### DISCUSSION THREAD

**Discussion by menny** [AnySoftKeyboard/AnySoftKeyboard#2655] [Ideas]
We will chat about https://github.com/AnySoftKeyboard/AnySoftKeyboard/pull/1870 and Gesture-Typing in general

**Comment by menny** [comment]
Hey, @EtienneDesticourt, do you want to give a quick update on what's the state of your PR?

**Reply by EtienneDesticourt** [reply]
Hey @menny, long time no see!
It should be about done. I rebased it with master this afternoon and added the final changes to the algorithm (all that was left to do was to divide the parameters by the key size so the algorithm could work for all keyboard sizes). I've added a bit of code to handle double letter words (pool/poll, to/too, put/putt, ...) while I was at it.

I've tested it on emulators of different sizes and types and on my physical device but that's not a lot of variety, so you're welcome to try it out. It should give the correct prediction instantly for 95% of words and hang half a second before giving a prediction for the last 5%.
If it hangs for every word as was reported by another user for a previous commit it means there is an issue and not enough words are being pruned.

**Reply by menny** [reply]
Wow, that's amazing! I'm going to pull your branch and check it out.

Do you think it is ready for production? I'm planning to release a new version sometime in the next week or so.

**Reply by EtienneDesticourt** [reply]
I don't know, I'm the only one who's tested it and the model parameters were derived from my writing patterns so I don't know how it's gonna behave on the full user population. I'd be more comfortable if we had a couple more people at least try it beforehand.

That being said it's been my daily driver for the past year and it's enough of an improvement over the previous iteration to make it worthy to get it out as soon as possible.

**Reply by Jonta** [reply]
Possible to make it an opt-in-setting at first perhaps? =)

**Reply by EtienneDesticourt** [reply]
It's already opt-in but if people are using the old one they're not going to be satisfied if it suddenly stops working for them.

**Comment by Z0pyrus** [comment]
I'm requesting a gesture feature:
I'm really looking forward to see the google gesture cursor control feature in AnySoft. Basically it's lets you use the space bar an you phone as a curser which is really useful. [(Here is a demonstration link)](https://youtu.be/V_4yMGKbe90?t=61)
I know there is already a beta setting (in Gesture Settings), which allows you to move the cursor around by swiping on the keyboard, but unfortunate, this will only works 1 character at a time and is quit annoying.

I really like your app!

**Comment by kocjs** [comment]
I would like to test and give feedback. Let me know how to assist. Do i just need to install the beta version?

**Comment by Jonta** [comment]
Can it weight user-added words more than ones in the default dictionary?

E.g. "I'm 55% sure that they swiped word X, which is in the default dictionary but that they never type, but I'm 45% sure they swiped word Y, which they added themselves, and type all the time. I'm going with Y on this one"

**Comment by MrXTheAnon** [comment]
I just wanna add another idea for swype, when you are adding a new word to the dictionary, the swype should auto detect the path of letters for that word.

**Reply by Jonta** [reply]
> when you are adding a new word to the dictionary, the swype should auto detect the path of letters for that word.

@MrXTheAnon When you are adding it, or after you've added it?

**Reply by MrXTheAnon** [reply]
After the word has been added to the dictionary.

**Reply by Jonta** [reply]
> After the word has been added to the dictionary.

@MrXTheAnon Does this happen to predefined dictionary words already? If not: Why should it with user-added words? =)

**Comment by SomeMichael** [comment]
Hi! I just tested swiping a little bit. And i must say it does quite a good job. Good work on that one!!!👍 it (but that i in "it" should be capitalized.😉)  works better than my Samsung original keyboard. 

One thing i would like to mention, though. When the algorithm chose the wrong word then i would like to be able to delete it with a single tap on the delete button. (Or is there a simple gesture to delete one word? 🤔)

Also while currently typing, the "I"s were not capitalized…

And short forms like "don't" and names like "Samsung" seam to be missing in the swipe db. And i can't swipe "missing"😆

Nonetheless, thanks a lot!

**Comment by Hooloovoo** [comment]
Many thanks to all for the hard work on this. I am using the existing gesture typing on ASK as my daily keyboard and am really happy to have a FOSS keyboard with some form of swype capability. 

I wondered if there is anything that we (the users) could be doing to help this effort. While I appreciate the privacy-first approach of ASK in general, I would opt-in to some level of data sharing and may spend some time swiping some prompts if it would help make this better. 

Out of interest, does gesture typing also use ASK's "word suggestions" as an input into the "best guess" of what someone swiped? I read https://github.com/AnySoftKeyboard/AnySoftKeyboard/pull/1870, but didn't see any mention of anything like that.

Similarly, do the best existing gesture keyboards use a language model as well? E.g. if I have already typed "The cat sat on the" then "mat" is probably more likely than "may", even if I finish slightly closer to "y" and "may" is more common in English usage than "mat".
